subsys/rpmsg_service/rpmsg_service.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020-2021, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <rpmsg_service/rpmsg_service.h>
+
+#include "rpmsg_backend.h"
+
+#include <zephyr.h>
+#include <device.h>
+#include <logging/log.h>
+
+#include <openamp/open_amp.h>
+
+#define LOG_LEVEL LOG_LEVEL_INFO
+#define LOG_MODULE_NAME rpmsg_service
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+/* Configuration defines */
+#if !DT_HAS_CHOSEN(zephyr_ipc_shm)
+#error "Module requires definition of shared memory for rpmsg"
+#endif
+
+#define MASTER IS_ENABLED(CONFIG_RPMSG_SERVICE_MODE_MASTER)
+
+static struct virtio_device vdev;
+static struct rpmsg_virtio_device rvdev;
+static struct metal_io_region *io;
+
+#if MASTER
+static struct rpmsg_virtio_shm_pool shpool;
+#endif
+
+static struct {
+	const char *name;
+	rpmsg_ept_cb cb;
+	struct rpmsg_endpoint ep;
+	volatile bool bound;
+} endpoints[CONFIG_RPMSG_SERVICE_NUM_ENDPOINTS];
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+#if MASTER
+
+static void ns_bind_cb(struct rpmsg_device *rdev,
```

Comment:
The name service announcement is not reserved for the master. 
But ok limit it at first, it will be extended to the slave later if necessary.